// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package iam;

message Access {
  repeated string members = 1;
  string resource = 2;
  string role = 3;
}

message GetAccessByPermissionsRequest {
  string role_name = 1;
  string permission_name = 2;
  bool expand_groups = 3;
  bool expand_resources = 4;
}

message GetAccessByResourcesRequest {
  string resource_name = 1;
  repeated string permission_names = 2;
  bool expand_groups = 3;
}

message GetAccessByResourcesReply {
  message Access {
    repeated string members = 1;
    string resource = 2;
    string role = 3;
  }
  repeated Access accesses = 1;
}

message DenormalizeRequest {
}

message DenormalizeReply {
	repeated AuthorizationTuple authorizations = 1;
}

message AuthorizationTuple {
	string member = 1;
	string resource = 2;
	string permission = 3;
}

message GetAccessByMembersRequest {
  string member_name = 1;
  repeated string permission_names = 2;
  bool expand_resources = 3;
}

message GetAccessByMembersReply {
  message Access {
    string member = 1;
    repeated string resources = 2;
    string role = 3;
  }
  repeated Access accesses = 1;
}

message PingRequest {
  string data = 1;
}

message PingReply {
  string data = 1;
}